topics/basics/testing_plugins/integration_tests/integration_tests_api.md

```diff
@@ -0,0 +1,197 @@
+
+The first parameter is the fully qualified name of the class that will correspond to the stub using the [`Remote`](%gh-ic%/platform/remote-driver/client/src/com/intellij/driver/client/Remote.kt) annotation.
+Strings are used to avoid introducing dependency between production and test code.
+
+The second parameter specifies `pluginId`, where classes are located.
+This parameter is required since IntelliJ-based IDEs use separate classloaders for each plugin, and the code that will call methods on the IDE side (Invoker) needs to know where to search for them.
+
+There is built-in support for `@Remote` annotation inside IntelliJ IDEA:
+![](remote-support.png)
+
+Rename and move the target class refactorings will update the annotation accordingly.
+Gutter icons can be used to navigate to the target class from a stub.
+
+## Calling IDE methods from tests
+
+The complete API test:
+
+```kotlin
+@Test
+fun testStubs() {
+  Starter.newContext(
+    "testExample",
+    TestCase(
+      IdeProductProvider.IC,
+      GitHubProject.fromGithub(
+        branchName = "master",
+        repoRelativeUrl = "JetBrains/ij-perf-report-aggregator"
+      )
+    )
+      .withVersion("2024.3")
+  )
+    .apply {
+      val pathToPlugin = System.getProperty("path.to.build.plugin")
+      PluginConfigurator(this).installPluginFromFolder(File(pathToPlugin))
+    }.runIdeWithDriver().useDriverAndCloseIde {
+      val storage = utility<PluginStorage>().getPluginStorage()
+      val key = storage.getKey()
+      val attributes = storage.getAttributes()
+      Assertions.assertEquals("static method", key)
+      Assertions.assertEquals(listOf("static1", "static2"), attributes)
+      val answer = service<PluginService>().getAnswer()
+      Assertions.assertEquals(42, answer)
+      waitForProjectOpen()
+      val project = singleProject()
+      val strings = service<PluginProjectService>(project).getStrings()
+      Assertions.assertArrayEquals(arrayOf("foo", "bar"), strings)
+    }
+}
+```
+
+There are two methods that allow invoking methods: `service` and `utility`.
+The first one will return an instance of service, and the second will return an instance of any class.
+
+The project-level service requires the `Project` stub.
+To get it, use the `singleProject` method, which is implemented in the same way as demonstrated above:
+
+```kotlin
+service<ProjectManager>().getOpenProjects().singleOrNull()
+```
+
+> Service and utility proxies can be acquired on each call, there is no need to cache them in clients.
+>
+{style="note"}
+
+## JMX/RMI limitations
+
+The main inconvenience of using JMX/RMI is that stubs need to be created for all objects which are used in tests.
+But on the other hand, they don't require modification of the production code.
+
+As with any protocol, JMX/RMI has its limitations:
+
+* Parameters and return values can only be:
+    * Primitives and their wrappers: `Integer`, `Short`, `Long`, `Double`, `Float`, and `Byte`.
```

Comment:
What about `Character`?

Reply:
No, characters are not supported at the moment. Using them will lead to:
```
Class class java.lang.Character is not annotated with @Remote annotation
```